lib/common/format.cc

```diff
@@ -0,0 +1,616 @@
+    case 'N': Advance(TokenKind::BN); break;
+    case 'Z': Advance(TokenKind::BZ); break;
+    default: token_.set_kind(TokenKind::B); break;
+    }
+    break;
+  case 'D':
+    switch (LookAheadChar()) {
+    case 'C': Advance(TokenKind::DC); break;
+    case 'P': Advance(TokenKind::DP); break;
+    case 'T': Advance(TokenKind::DT); break;
+    default: token_.set_kind(TokenKind::D); break;
+    }
+    break;
+  case 'E':
+    switch (LookAheadChar()) {
+    case 'N': Advance(TokenKind::EN); break;
+    case 'S': Advance(TokenKind::ES); break;
+    case 'X': Advance(TokenKind::EX); break;
+    default: token_.set_kind(TokenKind::E); break;
+    }
+    break;
+  case 'F': token_.set_kind(TokenKind::F); break;
+  case 'G': token_.set_kind(TokenKind::G); break;
+  case 'I': token_.set_kind(TokenKind::I); break;
+  case 'L': token_.set_kind(TokenKind::L); break;
+  case 'O': token_.set_kind(TokenKind::O); break;
+  case 'P': token_.set_kind(TokenKind::P); break;
+  case 'R':
+    switch (LookAheadChar()) {
+    case 'C': Advance(TokenKind::RC); break;
+    case 'D': Advance(TokenKind::RD); break;
+    case 'N': Advance(TokenKind::RN); break;
+    case 'P': Advance(TokenKind::RP); break;
+    case 'U': Advance(TokenKind::RU); break;
+    case 'Z': Advance(TokenKind::RZ); break;
+    default: token_.set_kind(TokenKind::None); break;
+    }
+    break;
+  case 'S':
+    switch (LookAheadChar()) {
+    case 'P': Advance(TokenKind::SP); break;
+    case 'S': Advance(TokenKind::SS); break;
+    default: token_.set_kind(TokenKind::S); break;
+    }
+    break;
+  case 'T':
+    switch (LookAheadChar()) {
+    case 'L': Advance(TokenKind::TL); break;
+    case 'R': Advance(TokenKind::TR); break;
+    default: token_.set_kind(TokenKind::T); break;
+    }
+    break;
+  case 'X': token_.set_kind(TokenKind::X); break;
+  case 'Z': token_.set_kind(TokenKind::Z); break;
+  case '-':
+  case '+': token_.set_kind(TokenKind::Sign); break;
+  case '/': token_.set_kind(TokenKind::Slash); break;
+  case '(': token_.set_kind(TokenKind::LParen); break;
+  case ')': token_.set_kind(TokenKind::RParen); break;
+  case '.': token_.set_kind(TokenKind::Point); break;
+  case ':': token_.set_kind(TokenKind::Colon); break;
+  case '\\': token_.set_kind(TokenKind::Backslash); break;
+  case '$': token_.set_kind(TokenKind::Dollar); break;
+  case '*':
+    token_.set_kind(LookAheadChar() == '(' ? TokenKind::Star : TokenKind::None);
+    break;
+  case ',': {
+    token_.set_kind(TokenKind::Comma);
+    CHAR laChar = LookAheadChar();
+    if (laChar == ',') {
+      Advance(TokenKind::Comma);
+      token_.set_offset(cursor_ - format_);
+      AppendError("Unexpected ',' in format expression");
+    } else if (laChar == ')') {
+      AppendError("Unexpected ',' before ')' in format expression");
+    }
+    break;
+  }
+  case '\'':
+  case '"':
+    for (++cursor_; cursor_ < end_; ++cursor_) {
+      if (*cursor_ == c) {
+        if (cursor_ >= end_ || *(cursor_ + 1) != c) {
```

Comment:
`cursor_` can never be >= `end_`  at this point in the body of the loop.

Reply:
Yes.  I caught that one as well.

Reply:
> `cursor_` can never be >= `end_` at this point in the body of the loop.

Maybe that should be `cursor_ == end_ -  1` because otherwise `*(cursor_ + 1)` accesses one past the end of the format in that case.

Also, `*(cursor_ + 1)` is usually written `cursor_[1]`.

Reply:
I will change to
        if (cursor_ + 1 < end_ && *(cursor_ + 1) != c) {
I prefer *(cursor_ + 1) as a companion to *cursor_ rather than cursor_[1], all the more so with the addition of the left side of the conjunction.

Reply:
`if (auto *nc{cursor_+1}; nc < end_ && *nc != c) { ... `

Reply:
That's a mix of elegance and .. je ne sais pas.  I'll go with it.